steps/container.py

```diff
@@ -233,3 +185,70 @@ def get_output(self, history=True):
     def remove_image(self, force=False):
         self.logging.info("Removing image %s" % self.image_id)
         d.remove_image(image=self.image_id, force=force)
+
+    def copy_file_to_container(self, src_file, dest_folder):
+        if not os.path.isabs(src_file):
+            src_file = os.path.abspath(os.path.join(os.getcwd(), src_file))
+
+        # The Docker library needs tar bytes to put_archive
+        f = tempfile.NamedTemporaryFile()
+        with tarfile.open(fileobj=f, mode='w') as t:
+            t.add(src_file, arcname=os.path.basename(src_file), recursive=False)
+
+        f.seek(0)
+
+        with open(f.name) as file:
+            d.put_archive(
+                container=self.container['Id'],
+                path=dest_folder,
+                data=file.read())
+
```

Comment:
We should close the file here: `f.close()`

Reply:
@goldmann I assumed that is what the 'with' part did. I'll reread the doc you pasted :)

Reply:
It's a bit more tricky here. It does, but when the file parameter is an actual file and not file object. If it would be closed by `with` statement, then you would not be able to read it again (to exec `put_archive`).

Reply:
Let me show an example when it would work, hold on.

Reply:
```python
tmp_file = tempfile.mktemp()

with tarfile.open(tmp_file, mode='w') as t:
    t.add(src_file, arcname=os.path.basename(src_file), recursive=False)

with open(tmp_file) as f:
    d.put_archive(
        container=self.container['Id'],
        path=dest_folder,
        data=f.read())

os.remove(tmp_file)
```

Reply:
Ah, I see: f vs file

Reply:
This works here
```
    def copy_file_to_container(self, src_file, dest_folder):
        if not os.path.isabs(src_file):
            src_file = os.path.abspath(os.path.join(os.getcwd(), src_file))

        # The Docker library needs tar bytes to put_archive
        with  tempfile.NamedTemporaryFile() as f:
            with tarfile.open(fileobj=f, mode='w') as t:
                t.add(src_file, arcname=os.path.basename(src_file), recursive=False)

            f.seek(0)

            with open(f.name) as file:
                d.put_archive(
                    container=self.container['Id'],
                    path=dest_folder,
                    data=file.read())
```
Any good?

Reply:
Or:

```python
with tempfile.NamedTemporaryFile() as f:
    with tarfile.open(fileobj=f, mode='w') as t:
        t.add(src_file, arcname=os.path.basename(src_file), recursive=False)

    f.seek(0)

    d.put_archive(
        container=self.container['Id'],
        path=dest_folder,
        data=file.read())
```

Reply:
Weird, I missed all your comments :)

Reply:
Your last one works fine, and is shorter than my attempt :)